### PR TITLE
[core] Mark `run_function_on_all_workers` as deprecated until we get rid of this.

### PR DESCRIPTION
**Pull request by fishbone**
Signed-off-by: Yi Cheng <74173148+iycheng@users.noreply.github.com>

## Why are these changes needed?
`run_function_on_all_workers` has a lot of issues:

- no guarantee that the function run before the remote function run.
- pubsub signal might be lost in some failure cases.

We can't delete it for now because the working dir init depends on this. So mark it as deprecated to warn people who uses this and delete it once we fix the working dir.

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by scv119** [comment]
do we know who is using it?

**Review by scv119 — APPROVED** [review]

File: python/ray/_private/worker.py [hunk 1]
```diff
@@ -683,6 +683,7 @@ def get_objects(self, object_refs: list, timeout: Optional[float] = None):
             debugger_breakpoint,
         )
 
+    @Deprecated
```
Comment (scv119) [inline python/ray/_private/worker.py:686]: can we add some comments explain why it's deprecated?  (as what you put in the PR summary)

Comment (iycheng) [reply python/ray/_private/worker.py:686]: Sure!

**Comment by iycheng** [comment]
@scv119  as far as I know, no one is using that. Modin has moved away from using this.

**Comment by rkooo567** [comment]
Does it make sense to deprecate it when this was never public APi?

**Comment by iycheng** [comment]
> Does it make sense to deprecate it when this was never public APi?

1. This is never public API. It's in ray/_private
2. Some developers might just use this. For example, data set tries to use this for customized SerDe.

I think this is just to double warn the people who want to go to the dangerous zones not to do this.

**Pull request merged**